examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+
+Each graph convolutional layer added captures information from a further level of neighbours.
+However, adding many graph convolutional layer can cause oversmoothing, where the model
+produces similar embeddings for all the nodes.
+
+Note that the `graph_info` passed to the constructor of the Keras model, and used as a *property*
+of the Keras model object, rather than input data for training or prediction.
+The model will accept a **batch** of `node_indices`, which are used to lookup the
+node features and neighbours from the `graph_info`.
+"""
+
+
+class GNNNodeClassifier(tf.keras.Model):
+    def __init__(
+        self,
+        graph_info,
+        num_classes,
+        hidden_units,
+        aggregation_type="sum",
+        combination_type="concat",
+        dropout_rate=0.2,
+        normalize=True,
+        *args,
+        **kwargs,
+    ):
+        super(GNNNodeClassifier, self).__init__(*args, **kwargs)
+
+        # Unpack graph_info to three elements: node_features, edges, and edge_weight.
+        node_features, edges, edge_weights = graph_info
+
+        self.node_features = node_features
+        self.edges = edges
+        self.edge_weights = edge_weights
+        # Set edge_weights to ones if not provided.
+        if self.edge_weights is None:
+            self.edge_weights = tf.ones(shape=edges.shape[1])
+        # Scale edge_weights to sum to 1.
+        self.edge_weights = self.edge_weights / tf.math.reduce_sum(self.edge_weights)
+
+        self.num_classes = num_classes
+        self.hidden_units = hidden_units
+        self.aggregation_type = aggregation_type
+        self.combination_type = combination_type
+        self.dropout_rate = dropout_rate
+        self.normalize = normalize
+
+    def build(self, input_shape):
+
+        self.preprocess = create_ffn(
+            self.hidden_units, self.dropout_rate, name="preprocess"
+        )
+
+        self.conv1 = GraphConvLayer(
+            self.hidden_units,
+            self.dropout_rate,
+            self.aggregation_type,
+            self.combination_type,
+            self.normalize,
+            name="graph_conv1",
+        )
+
+        self.conv2 = GraphConvLayer(
+            self.hidden_units,
+            self.dropout_rate,
+            self.aggregation_type,
+            self.combination_type,
+            self.normalize,
+            name="graph_conv2",
+        )
+
+        self.skip_connection = layers.Add(name="skip_connection")
+
+        self.postprocess = create_ffn(
+            self.hidden_units, self.dropout_rate, name="postprocess"
+        )
+
+        self.compute_logits = layers.Dense(units=self.num_classes, name="logits")
+
+    def call(self, input_node_indicies):
+        # Preprocess the node_features to produce node representations.
+        x = self.preprocess(self.node_features)
+        # Apply the first graph conv layer.
+        x1 = self.conv1((x, self.edges, self.edge_weights))
```

Comment:
I don't fully understand why edges and edge_weights are part of the call arguments if they don't vary from call to call. Sounds like they should be constructor arguments (making a layer instance specific to a set of edges)

Reply:
The reason I designed the `GraphConvLayer` to take node_features, edges, and weights as `call` parameters, rather than constructor parameters to allow using this `GraphConvLayer` in other models where you can pass a "subset" of the graph (nodes and edges) or a completely different graph in each train step. This can be useful in cases where the whole graph (info) cannot fit in memory, or in cases of graph level task (e.g. (sub)graph classification). 

In this example, we only have one graph (that can  fit in memory) and we want to preform a node-level task (node classification).  That is why in this particular example edges and edge_weights don't vary from one call to another.

WDYT?